if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)


setwd('~/Documents/ucl_postdoc/genotyping_sc/microarray/freemuxlet_microarray_correlation/')

data<-read.table('all_ref_clust_pearson_correlation.tsv', header = T)

rownames(data)<-data$Cluster
data2 <- as.matrix(data[,-1])

# Default Heatmap
heatmap(data2)

## select genotypes
freemux<-read.csv('../../all_samples_maternal_placenta_comb_freemux_object.csv',header = T)

genotypes<-unique(freemux[,c("sample","BEST.GUESS","genotype")])

genotypes2<-genotypes[grepl("Fetal|Maternal",genotypes$genotype),]

genotypes2$BEST.GUESS<-gsub(",.*","",genotypes2$BEST.GUESS)

genotypes2$sample_geno<-paste0(genotypes2$sample,"-",genotypes2$BEST.GUESS)
##### match with data
metadata<-data.frame(rownames(data2),gsub("CLUST","",rownames(data2)))

colnames(metadata)<-c("sample_clust","sample_geno")

metadata<-merge(metadata,genotypes2, by = "sample_geno", all.x = T, all.y = F)

######

data3<-as.data.frame(data2)
data3$sample_geno<-gsub("CLUST","",rownames(data2))
data3<-merge(data3,genotypes2, by = "sample_geno", all.x = T, all.y = F)
rownames(data3)<-data3$sample_geno


data4<-data3[,-c(1,18,19)]
colnames(data4)<-gsub(".*_","",colnames(data4))
baby<-colnames(data4)[grep("^B",colnames(data4))]
baby<-baby[order(baby)]

mum<-colnames(data4)[grep("^F",colnames(data4))]
mum<-mum[order(mum)]

data4<-data4[c(baby,mum,"genotype")]


col_split<-gsub("^B.*","Microarray Fetal",colnames(data4[,-17]))
col_split<-gsub("^F.*","Microarray Maternal",col_split)


##### add y markers expression: table generated by script 'extract_ychromosome_markers_sum...'
data4<-data4[!grepl('F1958NS-AB-sn|F1950PYI-AB-sn',rownames(data4)),]

data4$sample_id<-rownames(data4)


# change the name of the donor, to make it annomymised
tmp_names<-gsub("-.*","",rownames(data4))
tmp_names<- paste0("F", str_sub(tmp_names, start= -2))
tmp_names2 <- sub("^.*?-", "-", rownames(data4))
rownames(data4)<-paste0(tmp_names,tmp_names2)

tmp_names3<-str_split_fixed( rownames(data4),'-',3)[,c(1,2)]
tmp_names4<-paste0(tmp_names3[,1],"-",tmp_names3[,2],"-",substr(data4$genotype, 1, 3))
#non-unique values when setting 'row.names': ‘FCM-AB2-Fet’, ‘FCM-AB2-Mat’, ‘FVQ-AB2-Fet’, ‘FVQ-AB2-Mat’ 
tmp_names4[c(13,14)]<-gsub("AB2","AB2_2",tmp_names4[c(13,14)])
tmp_names4[c(3,4)]<-gsub("AB2","AB2_2",tmp_names4[c(3,4)])
rownames(data4)<-tmp_names4


# chage columns names
result <- paste0(substr(colnames(data4), 1, 1), substr(colnames(data4), nchar(colnames(data4)) - 1, nchar(colnames(data4))))
result[17] <-"genotype"
result[18] <-"sample_id"
colnames(data4) <-result



datmat<-data4[data4$genotype == "Maternal",]
datfet<-data4[data4$genotype == "Fetal",]
####
data5<-rbind(datmat,datfet)

row_split<-data5$genotype

metadata<-read.csv('~/Documents/ucl_postdoc/genotyping_sc/cellranger_v7/Aggr-FMI-All-GEX-TCR-aggregation-20230328-modified.csv')
#sex2<-data.frame(data5$sample_cl)
#colnames(sex2)[1]<-'sample_id'
#remove freemux group tag
#data5$sample_id<-substring(data5$sample_cl, 1, nchar(data5$sample_cl) - 2)
data5$sample_id<-result <- substr(data5$sample_id, 1, nchar(data5$sample_id) - 2)

data5$donor<-gsub("-.*","",data5$sample_id)
metadata$donor<-gsub("-.*","",metadata$sample_id)
fetalsex<-unique(metadata[,c("donor","Fetal.sex")])

#data5$Fetal.sex<-merge(data5,fetalsex,by = 'donor', all.x = T, all.y = FALSE)[,"Fetal.sex"]

# Create fetalsex dataframe, ensuring one sex per donor
fetalsex <- aggregate(Fetal.sex ~ donor, data = metadata, FUN = function(x) x[1])



merged_data<- left_join(data5, fetalsex, by = "donor")

rownames(merged_data)<-rownames(data5)



merged_data$Fetal.sex_long<-gsub("M","Male",merged_data$Fetal.sex)
merged_data$Fetal.sex_long<-gsub("F","Female",merged_data$Fetal.sex_long)



pdf("correlation_freemux_microarray_v3.pdf",width=6, height=7)
Heatmap(merged_data[,c(1:16)], name = "corr",split = row_split, column_split = col_split , 
        cluster_columns = FALSE,cluster_rows = FALSE,row_names_gp = gpar(fontsize = 7),right_annotation = rowAnnotation(B_sex = merged_data$Fetal.sex_long, 
                                                                                             col = list(B_sex = c("Male" = "orange", "Female" = "white"))))
dev.off()

### End of heatmap

###### plot freemuxlet groups features by tissue and genotype: 

samples_info<-read.csv('/Users/jose/Documents/ucl_postdoc/genotyping_sc/cellranger_v7/Aggr-FMI-All-GEX-TCR-aggregation-20230328-modified.csv', header = T)

names(freemux)[10]<-"sample_id"

freemux_sampleinfo<-merge(freemux,samples_info, by = "sample_id", all.x = T, all.y = F)

freemux_sampleinfo$genotype_tissue<-paste0(freemux_sampleinfo$genotype,"-",freemux_sampleinfo$Tissue)

#######
pdf("freemux_box_snps_genotypes.pdf",width=3.5, height=2)
p1 <- ggplot(freemux_sampleinfo_no_dbl, aes(x = NUM.SNPS, y = NUM.READS, col = genotype)) +
  geom_point(size = .4) +
  scale_color_manual("Genotype", values = c("Maternal" = "steelblue", "Fetal" = "burlywood1", "1,0" = "forestgreen")) +
  facet_grid( ~ genotype,scales = "free_y") + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
#scale_color_viridis_c('DIFF.LLK.BEST.NEXT',option = "C") + facet_grid(DROPLET.TYPE ~ genotype_tissue,scales = "free_y")
#geom_hline(yintercept=200, linetype="dashed", color = "blue") +
#geom_vline(xintercept=2500, linetype="dashed", color = "blue") + theme_minimal()
p1
dev.off()

### now barplots with proportions:
bp<-freemux_sampleinfo[grepl("Fetal|Maternal",freemux_sampleinfo$genotype),]
bp<-bp[grepl("SNG",bp$DROPLET.TYPE),]

### total nb of cells per sample:
bp1<-as.data.frame(table(bp$sample_id))
names(bp1)<-c('sample','n_cells')

### total of maternal samples
bp2<-as.data.frame(table(bp$sample_id,bp$Tissue,bp$Method,bp$genotype,bp$GA_Condition))
bp2<-bp2[bp2$Freq!=0,]
bp2$tissue

pdf('boxplot_n_cells_per_condition.pdf',width = 3.5, height = 3.5)
ggplot(bp2, aes(x=Var5, y=Freq, fill=Var4)) + 
  geom_boxplot(aes(fill=Var4)) + facet_wrap(~Var2, nrow = 2) + 
  ylab('n cells') + xlab ('G. age and condition') +
  scale_fill_manual("Genotype", values = c("Maternal" = "steelblue", "Fetal" = "burlywood1")) +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=6)) 
dev.off()
###### Cel type proportions: 

all_fmi<-readRDS(file = "/Users/jose/Documents/ucl_postdoc/FMI_supplementary_information/FMI-all-20patients-20240110.rds")


# This is the metadata
data <- all_fmi@meta.data

# Outpust of freemuxlet analysis are in cols:  data$genotype_two_libraries, data$Num.SNPs, data$Num.Reads


### need to run FMI_celltypes_ordered_list.R where cellgroups list is found

# Change the name of Lymphoid to T/NK
data$CellTypeManual.l1 <- ifelse(data$CellTypeManual.l1 == 'Lymphoid', 'T/NK',data$CellTypeManual.l1)

data$genotype_cor <- data$genotype_two_libraries
data$genotype_cor[data$Num.SNPs < 100]<-"Amb"
##################
##  Trophoblasts 
# Make a dataframe 
library(magrittr) # needs to be run every time you start R and want to use %>%
library(dplyr)  
t1 <- data %>% filter(CellTypeManual.l1 %in% "Trophoblast") %>%
  group_by(CellTypeManual.l3, genotype_cor) %>%
  summarise(n = n()) %>%
  mutate(proportion = n / sum(n)) %>%
  ungroup() 

t1 <- t1 %>% inner_join(cellgroups)

t1$cellgroup = factor(t1$cellgroup, levels = celltype.order.l1, ordered = TRUE)
t1$CellTypeManual.l3 = factor(t1$CellTypeManual.l3, levels = celltype_order.l3, ordered = TRUE)

# Plot
v1 <- t1 %>% ggplot(aes(x=CellTypeManual.l3, y=proportion, fill=genotype_cor)) +
  geom_col() + 
  # scale_fill_brewer(palette = "Paired") + "#66A61E", "#E6AB02"
  scale_fill_manual(values = c("#999999","#66A61E", "#E6AB02")) +
  geom_text(aes(label = n), position = position_stack(vjust = 0.5),size=2.5) + theme(axis.text.x=element_text(size=10, angle=90,hjust=0.95,vjust=0.2) ) 
# ggtitle(paste("Cell type proportion per patient: ", tissue, " - ", sep=""))

v1 <- v1 + coord_flip() + facet_grid(cellgroup ~ ., scales = "free", space = "free") +  theme(legend.position = "right")

v1
# ggsave(paste("/home/ssd/ysanchez/Projects/FMI-all-singlecell-20230308/outputs/singlecell-proportions-per-patient/barplot-proportion_Trophoblast_freemuxlet_per_cluster","_20240530.png",sep=''), v1, width=8,height=5, bg = "white")

##################
##  Fetal clusters
# Fetal clusters are those with more than 20% cells classified as fetal
fetal_clusters <- data %>% 
  group_by(CellTypeManual.l3, genotype_cor) %>%
  summarise(n = n()) %>%
  mutate(proportion = n / sum(n)) %>%
  filter(genotype_cor %in% "Fetal") %>%
  filter(proportion > 0.2) %>%
  ungroup() 

# add information about cell groups
# Add the information on cell groups
fetal_clusters <- fetal_clusters %>% inner_join(cellgroups)

# filter everything else that is not trophoblasts
fetal_clusters  <- fetal_clusters %>% filter(!cellgroup %in% "Trophoblast") 

# make a dataframe
e1 <- data %>% filter(CellTypeManual.l3 %in% fetal_clusters$CellTypeManual.l3) %>%
  group_by(CellTypeManual.l3, genotype_cor) %>%
  summarise(n = n()) %>%
  mutate(proportion = n / sum(n)) %>%
  ungroup() 

e1 <- e1 %>% inner_join(cellgroups)

e1$cellgroup = factor(e1$cellgroup, levels = celltype.order.l1, ordered = TRUE)
e1$CellTypeManual.l3 = factor(e1$CellTypeManual.l3, levels = celltype_order.l3, ordered = TRUE)

# Plot 
d1 <- e1 %>% ggplot(aes(x=CellTypeManual.l3, y=proportion, fill=genotype_cor)) +
  geom_col() + 
  # scale_fill_brewer(palette = "Paired") + "#66A61E", "#E6AB02"
  scale_fill_manual(values = c("#999999","#66A61E", "#E6AB02")) +
  geom_text(aes(label = n), position = position_stack(vjust = 0.5),size=2.5) + theme(axis.text.x=element_text(size=10, angle=90,hjust=0.95,vjust=0.2) ) 
# ggtitle(paste("Cell type proportion per patient: ", tissue, " - ", sep=""))

d1 <- d1 + coord_flip() + facet_grid(cellgroup ~ ., scales = "free", space = "free") +  theme(legend.position = "right")

d1

plot50<-v1 + d1

plot100<-v1 + d1

plotnothres<-v1 + d1

pdf("~/Documents/ucl_postdoc/FMI_supplementary_information/Trophoblast_genotypes_proportions.pdf",width=12, height=6)
plotnothres
plot50
plot100
dev.off()

